src/runtime/threadpool/builder.rs

```diff
@@ -330,14 +331,24 @@ impl Builder {
         // Get a handle to the clock for the runtime.
         let clock = self.clock.clone();
 
-        let pool = self.threadpool_builder
+        // Get the current trace dispatcher.
+        // TODO(eliza): when `tokio-trace-core` is stable enough to take a
+        // public API dependency, we should allow users to set a custom
+        // subscriber for the runtime.
+        let dispatch = dispatcher::get_default(Dispatch::clone);
+
+        let pool = self
+            .threadpool_builder
             .around_worker(move |w, enter| {
                 let index = w.id().to_usize();
+                let dispatch = dispatch.clone();
```

Comment:
Is this clone required? It seems like most of the other `with_default` things are fine with a reference.

Reply:
@seanmonstar A `Dispatch` is internally just an `Arc` --- we could potentially make `with_default` take an `&Dispatch` and perform the clone itself. I thought it was better to take it by value since the dispatcher might not always need to be cloned when `with_default` is used in other cases`...

Reply:
That’s a good point. with_dispatch should probably take a ref to follow the pattern. 

The fn can clone internally.

Reply:
I've opened https://github.com/tokio-rs/tokio/pull/971 to make this change. As I had suspected, it has the downside of requiring the arc to be cloned twice when entering a span using the `tokio-trace::Span` type. If we decide this is acceptable, I can rebase onto that branch.

Reply:
Setting the dispatcher every time a span is entered is unexpected to me. Can you expand on that decision?

Reply:
@carllerche 
> Setting the dispatcher every time a span is entered is unexpected to me. Can you expand on that decision?

Setting the default dispatcher on span entry is a relic of when spans tracked their parent's ID. At that time, it was necessary to ensure that any spans created inside a span were observed by the same subscriber that originally provided the entered span with an ID, as otherwise, new spans would be created with parent IDs that did not originate from that subscriber.

Now that spans don't track their parent ID, this is no longer necessary. However, removing this
behavior does mean that if a span is entered outside of the subscriber context it was created in,
any subsequent spans will be observed by the current default subscriber and thus will not be part of the original span's trace tree. Since subscribers are not expected to change frequently, and spans are not expected to move between them, this is likely acceptable.

I've removed this on my branch for #971 in commit 345567d90e6d8ab9937571f8e92ddf54bcf224d8. 

Not setting the dispatcher on entry does make span entry/exit significantly faster. Here are the results of running a benchmark that enters a span, does nothing, and immediately exits it, before this change:

```
test enter_span              ... bench:          93 ns/iter (+/- 14)
```

...and after:

```
test enter_span              ... bench:          51 ns/iter (+/- 9)
```